classes/plugins/PubIdPlugin.inc.php

```diff
@@ -32,333 +31,228 @@ function PubIdPlugin() {
+	 * @param $issue Issue
 	 */
-	function getPubObjectType($pubObject) {
-		$allowedTypes = array(
-			'Issue' => 'Issue',
-			'Article' => 'Article',
-			'ArticleGalley' => 'Galley',
-		);
-		$pubObjectType = null;
-		foreach ($allowedTypes as $allowedType => $pubObjectTypeCandidate) {
-			if (is_a($pubObject, $allowedType)) {
-				$pubObjectType = $pubObjectTypeCandidate;
-				break;
+	function excludeAndClearIssueObjects($exclude, $clear, $issue) {
+		$issueId = $issue->getId();
+		$articlePubIdEnabled = $this->isObjectTypeEnabled('Article', $issue->getJournalId());
+		$galleyPubIdEnabled = $this->isObjectTypeEnabled('ArticleGalley', $issue->getJournalId());
+		if (!$articlePubIdEnabled && !$galleyPubIdEnabled) return false;
+		$settingName = $this->getExcludeFormFieldName();
+		$pubIdType = $this->getPubIdType();
+		$articleDao =& DAORegistry::getDAO('ArticleDAO');
+		$publishedArticleDao =& DAORegistry::getDAO('PublishedArticleDAO');
+		$publishedArticles = $publishedArticleDao->getPublishedArticles($issueId);
+		foreach ($publishedArticles as $publishedArticle) {
+			if ($articlePubIdEnabled) {
+				if ($exclude && (!$publishedArticle->getStoredPubId($pubIdType) || $clear)) {
+					$publishedArticle->setData($settingName, 1);
+					$articleDao->updateObject($publishedArticle);
+				} else if ($clear) {
+					$articleDao->deletePubId($publishedArticle->getId(), $pubIdType);
+				}
+			}
+			if ($galleyPubIdEnabled) {
+				$articleGalleyDao =& DAORegistry::getDAO('ArticleGalleyDAO');
+				$articleGalleys =& $articleGalleyDao->getGalleysByArticle($publishedArticle->getId());
+				foreach ($articleGalleys as $articleGalley) {
+					if ($exclude && (!$articleGalley->getStoredPubId($pubIdType) || $clear)) {
+						$articleGalley->setData($settingName, 1);
+						$articleGalleyDao->updateObject($articleGalley);
+					} else if ($clear) {
+						$articleGalleyDao->deletePubId($articleGalley->getId(), $pubIdType);
+					}
+				}
 			}
 		}
-		if (is_null($pubObjectType)) {
-			// This must be a dev error, so bail with an assertion.
-			assert(false);
-			return null;
-		}
-		return $pubObjectType;
-	}
-
-	/**
-	 * Set and store a public identifier.
-	 * @param $pubObject Issue|Article|ArticleGalley
-	 * @param $pubObjectType string As returned from self::getPubObjectType()
-	 * @param $pubId string
-	 * @return string
-	 */
-	function setStoredPubId(&$pubObject, $pubObjectType, $pubId) {
-		$dao = $this->getDAO($pubObjectType);
-		$dao->changePubId($pubObject->getId(), $this->getPubIdType(), $pubId);
-		$pubObject->setStoredPubId($this->getPubIdType(), $pubId);
 	}
 
 	/**
-	 * Return the name of the corresponding DAO.
-	 * @param $pubObject object
-	 * @return DAO
+	 * @copydoc PKPPubIdPlugin::getDAOs()
 	 */
-	function getDAO($pubObjectType) {
-		$daos =  array(
+	function getDAOs() {
+		return  array(
 			'Issue' => 'IssueDAO',
 			'Article' => 'ArticleDAO',
-			'Galley' => 'ArticleGalleyDAO',
+			'ArticleGalley' => 'ArticleGalleyDAO',
```

Comment:
(Note that there are Representation classes, and Application::getRepresentationDAO, etc., so it might be possible to move some of these concerns into PKP-lib and then extend/subclass. That'll mean making static helper classes instantiations instead.)